### PR TITLE
53350 Mock Auth USiP configurations

### PR DESCRIPTION
**Pull request by acaruso-oddball**
## Summary

This PR updates the configuration of the Unified Sign in Page to include `vmack`, a new application to mock authentication from a CSP standpoint. This is achieved through the addition of the `mockLogin` utility function. When used this function redirects the end user (engineers, etc.) to the intended mock authentication url: `<environment>/v0/sign_in/authorize?client_id=vamock` 

## Related issue(s)
Related Issue: [[FE] (Mock Authentication) - Add vamock to USiP configurations #53350](https://github.com/department-of-veterans-affairs/va.gov-team/issues/53350)
Related Epic: [Mocked Authentication for Localhost and Development stacks](https://github.com/department-of-veterans-affairs/va.gov-team/issues/53015)

## Testing done
Mock Authentication is a new feature for the platform. As this PR primarily focuses on adding rather than updating functionality to the authentication process, it does not adversely impact . Unit tests were added to ensure updated worked as intended. In order to manually test these changes, one could add the mockLogin utility function to the onclick of a button. 

## What areas of the site does it impact?
The mock authentication flow.

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature
- [ ]  [Accessibility foundational testing](https://depo-platform-documentation.scrollhelp.site/developer-docs/wcag-2-1-success-criteria-and-foundational-testing) has been performed

### CODE REVIEW

File: src/platform/user/authentication/config/staging.config.js [hunk 1]
```diff
@@ -9,8 +9,12 @@ import {
...
 export default {
   default: {
-    allowedSignInProviders: { ...defaultSignInProviders },
-    allowedSignUpProviders: { ...defaultSignUpProviders },
+    allowedSignInProviders: {
+      ...defaultSignInProviders,
+    },
+    allowedSignUpProviders: {
+      ...defaultSignUpProviders,
+    },
```
Comment (asg5704) [inline src/platform/user/authentication/config/staging.config.js:17]: The `vamock` functionality is only for dev & localhost, staging changes not required.

File: src/platform/user/authentication/utilities.js [hunk 2]
```diff
@@ -285,18 +286,21 @@ export function redirect(redirectUrl, clickedEvent, type = '') {
   window.location = redirectUrl;
 }
 
+export async function mockLogin({ clickedEvent = AUTH_EVENTS.MOCK_LOGIN }) {
+  const url = `${MOCK_LOGIN_URL}`;
```
Comment (asg5704) [inline src/platform/user/authentication/utilities.js:290]: Generate the url using the `createOAuthRequest`, it should still have the other OAuth query params besides the `type`.

File: src/platform/user/authentication/constants.js [hunk 3]
```diff
@@ -11,10 +11,15 @@ import {
 export const API_VERSION = 'v1';
 export const FORCE_NEEDED = 'force-needed';
 
+export const MOCK_LOGIN_URL = `${
+  environment.API_URL
+}/v0/sign_in/authorize?client_id=vamock`;
+
```
Comment (asg5704) [inline src/platform/user/authentication/constants.js:17]: Remove and see note about `mockLogin`

File: src/platform/user/authentication/config/staging.config.js (hunk 1 above)
Comment (acaruso-oddball) [reply src/platform/user/authentication/config/staging.config.js:17]: Duly noted. If you look at the commits over time you see that changes made to this file were reverted back to its original functionality and what is highlighted above is the same as before, I will update this file to remove extra lines created.

**Review by asg5704 — CHANGES_REQUESTED** [review]
Looks mostly good at first pass, still needs a few tweaks.

File: src/platform/user/authentication/utilities.js (hunk 2 above)
Comment (acaruso-oddball) [reply src/platform/user/authentication/utilities.js:290]: Sounds good, will make these adjustments.

File: src/platform/user/authentication/config/staging.config.js (hunk 1 above)
Comment (acaruso-oddball) [reply src/platform/user/authentication/config/staging.config.js]: Hi Alex, I've made these changes.

File: src/platform/user/authentication/constants.js (hunk 3 above)
Comment (acaruso-oddball) [reply src/platform/user/authentication/constants.js]: Hi Alex, I've made these changes.

File: src/platform/utilities/oauth/utilities.js [hunk 4]
```diff
@@ -96,11 +96,15 @@ export async function createOAuthRequest({
...
+    !application ||
+    [CLIENT_IDS.WEB, CLIENT_IDS.VAWEB, CLIENT_IDS.VAMOCK].includes(clientId);
   const isMobileOAuth =
     [EXTERNAL_APPS.VA_FLAGSHIP_MOBILE, EXTERNAL_APPS.VA_OCC_MOBILE].includes(
       application,
-    ) || [CLIENT_IDS.MOBILE, CLIENT_IDS.VAMOBILE].includes(clientId);
+    ) ||
+    [CLIENT_IDS.MOBILE, CLIENT_IDS.VAMOBILE, CLIENT_IDS.VAMOCK].includes(
+      clientId,
+    );
```
Comment (asg5704) [inline src/platform/utilities/oauth/utilities.js:107]: 🙅

File: src/platform/utilities/oauth/utilities.js [hunk 5]
```diff
@@ -96,11 +96,15 @@ export async function createOAuthRequest({
   acr,
 }) {
   const isDefaultOAuth =
-    !application || [CLIENT_IDS.WEB, CLIENT_IDS.VAWEB].includes(clientId);
+    !application ||
+    [CLIENT_IDS.WEB, CLIENT_IDS.VAWEB, CLIENT_IDS.VAMOCK].includes(clientId);
```
Comment (asg5704) [inline src/platform/utilities/oauth/utilities.js:100]: 🚀

File: src/platform/user/authentication/utilities.js [hunk 6]
```diff
@@ -285,18 +285,41 @@ export function redirect(redirectUrl, clickedEvent, type = '') {
   window.location = redirectUrl;
 }
 
+export async function mockLogin({ clickedEvent = AUTH_EVENTS.MOCK_LOGIN }) {
+  const { application, codeChallenge, codeChallengeMethod } = getQueryParams();
+  const gaClientId = '';
+  const config =
+    externalApplicationsConfig[application] ||
+    externalApplicationsConfig.default;
```
Comment (asg5704) [inline src/platform/user/authentication/utilities.js:293]: ? Can you explain why this is needed?

**Comment by acaruso-oddball** [comment]
@asg5704 updates complete and tests passing.

File: src/platform/user/authentication/utilities.js [hunk 7]
```diff
@@ -167,8 +167,9 @@ export function sessionTypeUrl({
   allowVerification = false,
   useOauth = false,
   acr = null,
+  mock,
 }) {
-  if (!type) {
+  if (!type && !mock) {
```
[the remark this answers is not in the document]
Comment (acaruso-oddball) [reply src/platform/user/authentication/utilities.js:172]: 👍🏾

File: src/platform/utilities/oauth/utilities.js (hunk 4 above)
Comment (acaruso-oddball) [reply src/platform/utilities/oauth/utilities.js:107]: 👍🏾

File: src/platform/utilities/oauth/utilities.js (hunk 5 above)
Comment (acaruso-oddball) [reply src/platform/utilities/oauth/utilities.js:100]: 👍🏾

File: src/platform/user/tests/authentication/utilities.unit.spec.js [hunk 8]
```diff
@@ -435,6 +435,19 @@ describe('Authentication Utilities', () => {
     });
   });
 
+  describe('mockLogin', () => {
+    it('should redirect to proper mockLogin url', async () => {
+      setup({
+        path: usipPath,
+      });
+      await authUtilities.mockLogin({ isOAuth: true });
```
[the remark this answers is not in the document]
Comment (acaruso-oddball) [reply src/platform/user/tests/authentication/utilities.unit.spec.js:443]: used to mirror login test but upon removal tests still pass so will remove.

File: src/platform/user/authentication/utilities.js (hunk 6 above)
Comment (acaruso-oddball) [reply src/platform/user/authentication/utilities.js]: Was added to mirror code throughout this section, but upon testing it's deemed unnecessary so removed.

**Review by asg5704 — APPROVED** [review]

**Comment by acaruso-oddball** [comment]
thx @asg5704 Alex

**Comment by acaruso-oddball** [comment]
Hi @department-of-veterans-affairs/va-platform-cop-frontend can I get a review?

**Review by rmessina1010 — APPROVED** [review]
LGTM

**Pull request merged**